### PR TITLE
HIVE-26293: Migrate remaining exclusive DDL operations to EXCL_WRITE lock & bug fixes

### PR DESCRIPTION
**Pull request by deniskuzZ**
### What changes were proposed in this pull request?

Migrate operations that still require an EXCLUSIVE lock to EXCL_WRITE. 

### Why are the changes needed?

Part of lockless-reads story

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/misc/msck/MsckAnalyzer.java [hunk 1]
```diff
@@ -87,7 +87,8 @@ public void analyzeInternal(ASTNode root) throws SemanticException {
     }
 
     if (repair && AcidUtils.isTransactionalTable(table)) {
-      outputs.add(new WriteEntity(table, WriteType.DDL_EXCLUSIVE));
+      outputs.add(new WriteEntity(table, AcidUtils.isLocklessReadsEnabled(table, conf) ?
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/misc/msck/MsckAnalyzer.java:90]: Maybe a little late, maybe it is not viable, but I feel that we spread this checks all around the code now. Would it be possible to translate the `WriteType` to the lock type in one place instead?

Maybe in a different PR later?

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 2]
```diff
@@ -2429,7 +2433,7 @@ private void testConcurrent2MergeUpdatesConflict(boolean slowCompile) throws Exc
     List<String> res = new ArrayList<>();
     driver.getFetchTask().fetch(res);
     Assert.assertEquals(2, res.size());
-    Assert.assertEquals("Lost Update", "[earl\t10, amy\t10]", res.toString());
+    Assert.assertTrue("Lost Update", isEqualCollection(res, asList("earl\t10", "amy\t10")));
```
Comment (pvary) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:2436]: nit: Why not `Assert.assertArrayEquals`?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/AcidEventListener.java [hunk 3]
```diff
@@ -80,22 +82,18 @@ public void onDropTable(DropTableEvent tableEvent) throws MetaException {
...
-        long currentTxn = Optional.ofNullable(tableEvent.getEnvironmentContext())
-          .map(EnvironmentContext::getProperties)
-          .map(prop -> prop.get("txnId"))
-          .map(Long::parseLong)
-          .orElse(0L);
+        long currentTxn = getTxnId(tableEvent.getEnvironmentContext());
         
-        try {
-          if (currentTxn > 0) {
+        if (currentTxn > 0) {
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/AcidEventListener.java:87]: nit: Why is this change?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/misc/msck/MsckAnalyzer.java (hunk 1 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/misc/msck/MsckAnalyzer.java:90]: probably we could try to make this translation in AcidUtils.makeLockComponents, however, there might be many hidden stones, better to try it in diff PR

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java (hunk 2 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:2436]: it uses Arrays.deepEquals that is order sensitive, unlike the isEqualCollection

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/AcidEventListener.java (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/AcidEventListener.java:87]: just a small refactor, moved if out of try block

**Comment by deniskuzZ** [comment]
> @deniskuzZ: Would it worth to separate out the bugfixes to another jira? If a customer needs a fix, it would be easier later to just provide the fix without the whole change

for the lockless-reads feature, the whole stuff should go in, that is why I decided to keep it here

**Review by pvary — APPROVED** [review]

**Pull request merged**